util/include/MultiSizeBufferPool.hpp

```diff
@@ -0,0 +1,498 @@
+ * and restrict the total pool size.
+ * - Rich statistics:
+ * 1) Histograms (snapshots) reported to logs. Currently, histograms are for this pool can't be viewed
+ * from concord-ctl.
+ * 2) Metrics - reported to Wavefront.
+ * 3) Statistics - reported to log. It is very important to mention, that due to speed consideration, each of the
+ * statistics are "stand-alone" and atomic. That means, for a statistics report, you might find discrepancies between
+ * statistics of the same type, due to the nature of this multi-threaded pool (statistics are updated all the time by
+ * multiple clients in a non-atomic way). All types of statistics can be disabled/enabled.
+ * - Blocking - calls to get a buffer should never fail, unless requested buffer size is too big or there is no more
+ * memory - in rare cases user may get a larger buffer than requested, in more cases caller may block.
+ * Currently, there is no implementation for wait_for(..) e.g async mode.
+ * - Dynamic allocation - each pool allocates its initial chunks when created. In case there is no available
+ * chunks in a specific subpool, a new one gets dynamically allocated - up to a maximum number of chunks defined for
+ * this subpool.
+ * - Throwing - asserts are done to detect internal coding errors. In case of user input error or System error -
+ * an exception is thrown.
+ * - Borrowing - if all buffers of a certain sub-pool SP1 are allocated, a calling thread checks for a free chunk in the
+ * rest of the subpools (SP2, SP3 etc..) for speed, but won't try to allocate a new one on these subpools. If a buffer
+ * isn't found - it returns to wait for a chunk to be released by another thread on the original subpool SP1.
+ * - Unit-testing: done by GTest, see MultiSizeBufferPool_test.
+ * - Safety checks:
+ * 1) Pool assign magic flags on chunk's header and trailer. If they are changed - exception is thrown.
+ * 2) Check for leaks - pool can be configured to check that all buffers have been returned to it before destroying
+ * itself. If not, it throws.
+ *
+ * Locking is done only for conditional variables for this pool. This is in order to achieve speed. This pool currently
+ * faces thousands of calls per second for its main use-case. This leads to sub-optimized cases in both statistics and
+ * behavior. For example, the upper threshold for maximal pool size might cross more than user defined, due to the
+ * multi-threaded nature and since there is no locking. Another example is for purge limit: we might purge a "little
+ * more" than requested.
+ *
+ * Important comment about usage:
+ * It is the responsibilty iof the user to make sure that:
+ * 1) Pool is destroyed by a single thread
+ * 2) whn pool is destroyed all buffers have been returned
+ * 3) This implies that since all buffers have been returned, no more threads are waiting for buffers.
+ */
+
+namespace concordUtil {
+
+namespace test {
+class MultiSizeBufferPoolTestFixture;
+}
+
+class MultiSizeBufferPool {
```

Comment:
Gil, please introduce an interface for Pool.

Reply:
There is already this line at the top of the header comments at the beggining of the file.
There is a continuation task to remove some of the structs here and transform to impl, but I cant do it now (log crashes, no time to deal with it, had to revert)
see:
`=== For API: search for "Public API" ===`